File: src/api/Shift.Api/Platform/DashboardController.cs
using Microsoft.AspNetCore.Mvc;

using Shift.Sdk.Service.Platform;
using Shift.Sdk.Service.Platform.DashboardNotifications;

namespace Shift.Api.Platform;

[ApiController]
[ApiExplorerSettings(GroupName = "Platform API: Dashboard")]
public class DashboardController(IDashboardNotificationManager notificationManager) : ShiftControllerBase
{
    public class SearchResult
    {
        public required DashboardNotification[] Notifications { get; init; }
        public required Dictionary<Guid, string> Users { get; init; }
        public required Dictionary<Guid, bool> VisibleOnDashboard { get; init; }
    }

    [HttpGet("api/platform/dashboard")]
    [HybridPermission("platform/dashboard", DataAccess.Read)]
    [EndpointName("retrieveDashboard")]
    public async Task<ActionResult<DashboardData>> RetrieveDashboardAsync(IPrincipalProvider principalProvider, IDashboardService service)
    {
        var dashboard = await service.CreateAsync(principalProvider.OrganizationId);

        return Ok(dashboard);
    }

    [HttpPost("api/platform/dashboard/notifications/search")]
    [HybridPermission("platform/dashboard-notifications", DataAccess.Read)]
    [EndpointName("searchNotifications")]
    public async Task<ActionResult<SearchResult>> SearchNotificationsAsync(UserReader userReader, DashboardNotificationCriteria  criteria)
    {
        if (criteria == null)
            criteria = new DashboardNotificationCriteria();

        if (criteria.Filter == null)
            criteria.Filter = new QueryFilter();
        else
            criteria.Filter.PageSize = new QueryFilter().PageSize;

        var result = notificationManager.SearchNotifications(criteria);

        Dictionary<Guid, string> users;

        if (result.Notifications.Length > 0)
        {
            var userCriteria = new SearchUsers
            {
                UserIds = result.Notifications
                    .Select(x => x.CreatedBy)
                    .Union(result.Notifications.Select(x => x.ModifiedBy))
                    .Distinct().ToArray()
            };
            userCriteria.DisablePaging();

            users = (await userReader.SearchAsync(userCriteria)).ToDictionary(x => x.UserId, x => x.FullName);
        }
        else
            users = new Dictionary<Guid, string>();

        Response.AddPagination(criteria.Filter, result.TotalCount);

        return Ok(new SearchResult
        {
            Notifications = result.Notifications,
            Users = users,
            VisibleOnDashboard = result.VisibleOnDashboard
        });
    }

    [HttpGet("api/platform/dashboard/notifications/{notificationId}")]
    [HybridPermission("platform/dashboard-notifications", DataAccess.Read)]
    [ProducesResponseType<DashboardNotification>(StatusCodes.Status200OK)]
    [EndpointName("retrieveNotification")]
    public IActionResult RetrieveNotification(Guid notificationId)
    {
        var notification = notificationManager.RetrieveNotification(notificationId);
        return Ok(notification);
    }

    [HttpPost("api/platform/dashboard/notifications")]
    [HybridPermission("platform/dashboard-notifications", DataAccess.Update)]
    [ProducesResponseType<DashboardNotification>(StatusCodes.Status200OK)]
    [EndpointName("modifyNotification")]
    public IActionResult ModifyNotification(IPrincipalProvider principalProvider, DashboardNotification notification)
    {
        var principal = principalProvider.GetPrincipal();

        notificationManager.ModifyNotification(notification, principal.UserId);

        return Ok(notification);
    }

    [HttpDelete("api/platform/dashboard/notifications/{notificationId}")]
    [HybridPermission("platform/dashboard-notifications", DataAccess.Delete)]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [EndpointName("deleteNotification")]
    public IActionResult DeleteNotification(Guid notificationId)
    {
        notificationManager.DeleteNotification(notificationId);

        return Ok(new {});
    }
}

File: src/ui/Shift.UI/src/api/controllers/platform/dashboard/_dashboardController.ts
import { fetchHelper } from "@/api/fetchHelper";
import { ApiDashboardNotificationCriteria } from "./ApiDashboardNotificationCriteria";
import { QueryResult } from "@/models/QueryResult";
import { ApiDashboardNotification } from "./ApiDashboardNotification";
import { ApiDashboardSearchResult } from "./ApiDashboardSearchResult";
import { ApiDashboardNotificationType } from "./ApiDashboardNotificationType";
import { ApiDashboard } from "./ApiDashboard";

interface ModifiedNotification {
    NotificationId: string | null;
    Type: ApiDashboardNotificationType;
    Title: string;
    Details: string | null;
    LinkText: string | null;
    LinkUrl: string | null;
    StartDate: string | null;
    EndDate: string | null;
    IsActive: boolean;
}

export const _dashboardController = {
    async retrieveDashboard(): Promise<ApiDashboard | null> {
        return await fetchHelper.get<ApiDashboard>("/api/platform/dashboard");
    },

    async searchNotifications(query: ApiDashboardNotificationCriteria, pageIndex: number): Promise<{
        queryResult: QueryResult<ApiDashboardNotification>;
        users: Record<string, string>,
        visibleOnDashboard: Record<string, boolean>,
    } | null>
    {
        const result = await fetchHelper.getPagedRows<object>(`/api/platform/dashboard/notifications/search`, query, pageIndex, null, null, null);
        if (!result) {
            return null;
        }

        const searchResult = result.rows as unknown as ApiDashboardSearchResult;

        return {
            queryResult: {
                pageIndex: result.pageIndex,
                rows: searchResult.Notifications,
                totalRowCount: result.totalRowCount,
                rowsPerPage: result.rowsPerPage
            },
            users: searchResult.Users,
            visibleOnDashboard: searchResult.VisibleOnDashboard,
        }
    },

    async retrieveNotification(notificationId: string): Promise<ApiDashboardNotification | null> {
        return await fetchHelper.get(`/api/platform/dashboard/notifications/${notificationId}`);
    },

    async modifyNotification(notification: ModifiedNotification): Promise<ApiDashboardNotification | null> {
        return await fetchHelper.post("/api/platform/dashboard/notifications", notification);
    },

    async deleteNotification(notificationId: string): Promise<void> {
        await fetchHelper.delete(`/api/platform/dashboard/notifications/${notificationId}`, null);
    },
}